File: 05. Tree/[429]N 叉树的层序遍历.java
//给定一个 N 叉树，返回其节点值的层序遍历。（即从左到右，逐层遍历）。 
//
// 树的序列化输入是用层序遍历，每组子节点都由 null 值分隔（参见示例）。 
//
// 
//
// 示例 1： 
//
// 
//
// 
//输入：root = [1,null,3,2,4,null,5,6]
//输出：[[1],[3,2,4],[5,6]]
// 
//
// 示例 2： 
//
// 
//
// 
//输入：root = [1,null,2,3,4,5,null,null,6,7,null,8,null,9,10,null,null,11,null,12,
//null,13,null,null,14]
//输出：[[1],[2,3,4,5],[6,7,8,9,10],[11,12,13],[14]]
// 
//
// 
//
// 提示： 
//
// 
// 树的高度不会超过 1000 
// 树的节点总数在 [0, 10^4] 之间 
// 
// Related Topics 树 广度优先搜索 
// 👍 148 👎 0


//leetcode submit region begin(Prohibit modification and deletion)
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
    public List<List<Integer>> levelOrder(Node root) {
        Queue<Node> Q = new LinkedList<>();
        List<List<Integer>> ans = new ArrayList<>();

        if (root != null) {
            Q.offer(root);
        }
        while (Q.size() > 0) {
            final int qSize = Q.size();
            List<Integer> tmp = new ArrayList<>();
            for (int i = 0; i < qSize; i++) {
                Node cur = Q.poll();
                tmp.add(cur.val);
                for (Node child : cur.children) {
                    if (child != null) {
                        Q.offer(child);
                    }
                }
            }
            ans.add(tmp);
        }
        return ans;
    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: 02. Queue/[429]N 叉树的层序遍历.py
# 给定一个 N 叉树，返回其节点值的层序遍历。（即从左到右，逐层遍历）。 
# 
#  树的序列化输入是用层序遍历，每组子节点都由 null 值分隔（参见示例）。 
# 
#  
# 
#  示例 1： 
# 
#  
# 
#  
# 输入：root = [1,null,3,2,4,null,5,6]
# 输出：[[1],[3,2,4],[5,6]]
#  
# 
#  示例 2： 
# 
#  
# 
#  
# 输入：root = [1,null,2,3,4,5,null,null,6,7,null,8,null,9,10,null,null,11,null,12,
# null,13,null,null,14]
# 输出：[[1],[2,3,4,5],[6,7,8,9,10],[11,12,13],[14]]
#  
# 
#  
# 
#  提示： 
# 
#  
#  树的高度不会超过 1000 
#  树的节点总数在 [0, 10^4] 之间 
#  
# 
#  Related Topics 树 广度优先搜索 👍 337 👎 0


# leetcode submit region begin(Prohibit modification and deletion)
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""
class Solution(object):
    def levelOrder(self, root):
        """
        :type root: Node
        :rtype: List[List[int]]
        """
        cur_level = list()
        ans = list()
        if root:
            cur_level.append(root)

        while len(cur_level)>0:
            next_level = list()
            cur_level_val = list()
            for cur_node in cur_level:
                cur_level_val.append(cur_node.val)
                if cur_node.children:
                    next_level.extend(cur_node.children)

            cur_level = next_level
            ans.append(cur_level_val)
        return ans
        
# leetcode submit region end(Prohibit modification and deletion)
